plugin/federation/federation.go

```diff
@@ -243,100 +246,102 @@ func (f *federation) setEntities(schema *ast.Schema) {
+					if len(dir.Arguments) != 1 || dir.Arguments[0].Name != "fields" {
+						panic("Exactly one `fields` argument needed for @key declaration.")
 					}
+					arg := dir.Arguments[0]
+					keyFieldSet := fieldset.New(arg.Value.Raw, nil)
 
-					keyFields[i] = &KeyField{Definition: def, Field: field}
-				}
-
-				e := &Entity{
-					Name:      schemaType.Name,
-					KeyFields: keyFields,
-					Def:       schemaType,
-					Requires:  requires,
-				}
+					keyFields := make([]*KeyField, len(keyFieldSet))
+					resolverFields := []string{}
+					for i, field := range keyFieldSet {
+						def := field.FieldDefinition(schemaType, schema)
 
-				// Let's process custom entity resolver settings.
-				dir = schemaType.Directives.ForName("entityResolver")
-				if dir != nil {
-					if dirArg := dir.Arguments.ForName("multi"); dirArg != nil {
-						if dirVal, err := dirArg.Value.Value(nil); err == nil {
-							e.Multi = dirVal.(bool)
+						if def == nil {
+							panic(fmt.Sprintf("no field for %v", field))
 						}
-					}
-				}
 
-				// If our schema has a field with a type defined in
-				// another service, then we need to define an "empty
-				// extend" of that type in this service, so this service
-				// knows what the type is like.  But the graphql-server
-				// will never ask us to actually resolve this "empty
-				// extend", so we don't require a resolver function for
-				// it.  (Well, it will never ask in practice; it's
-				// unclear whether the spec guarantees this.  See
-				// https://github.com/apollographql/apollo-server/issues/3852
-				// ).  Example:
-				//    type MyType {
-				//       myvar: TypeDefinedInOtherService
-				//    }
-				//    // Federation needs this type, but
-				//    // it doesn't need a resolver for it!
-				//    extend TypeDefinedInOtherService @key(fields: "id") {
-				//       id: ID @external
-				//    }
-				if !e.allFieldsAreExternal() {
-					resolverFields := []string{}
-					for _, f := range e.KeyFields {
-						resolverFields = append(resolverFields, f.Field.ToGo())
+						keyFields[i] = &KeyField{Definition: def, Field: field}
+						resolverFields = append(resolverFields, keyFields[i].Field.ToGo())
 					}
 
 					resolverFieldsToGo := schemaType.Name + "By" + strings.Join(resolverFields, "And")
+					var resolverName string
 					if e.Multi {
 						resolverFieldsToGo += "s" // Pluralize for better API readability
-						e.ResolverName = fmt.Sprintf("findMany%s", resolverFieldsToGo)
+						resolverName = fmt.Sprintf("findMany%s", resolverFieldsToGo)
 					} else {
-						e.ResolverName = fmt.Sprintf("find%s", resolverFieldsToGo)
+						resolverName = fmt.Sprintf("find%s", resolverFieldsToGo)
 					}
 
-					e.InputType = resolverFieldsToGo + "Input"
+					e.Resolvers = append(e.Resolvers, &EntityResolver{
+						ResolverName: resolverName,
+						KeyFields:    keyFields,
+						InputType:    resolverFieldsToGo + "Input",
+					})
+
+					e.Requires = []*Requires{}
```

Comment:
I think we want to process `requires` directives outside of the loop for processing all the `keys` to avoid reprocessing them.

Reply:
Yeah, dumb oversight on my part!

Reply:
all good! these changes were very tricky.  and hard to catch in unit tests since the generated should still work correctly. 🤷